reconcile/ocm_machine_pools.py

```diff
@@ -1,137 +1,400 @@
+            labels=pool.labels,
+            subnet=pool.subnet,
+            cluster=cluster,
+        )
+
+
+class PoolHandler(BaseModel):
+    # Class, that acts on a pool, based on the action
+
+    action: str
+    pool: AbstractPool
+
+    def act(self, dry_run: bool, ocm: OCM) -> None:
+        logging.info(f"{self.action} {self.pool.dict(by_alias=True)}")
+        if dry_run:
+            return
+
+        if not self.action:
+            pass
+        elif self.action == "delete":
+            self.pool.delete(ocm)
+        elif self.action == "create":
+            self.pool.create(ocm)
+        elif self.action == "update":
+            self.pool.update(ocm)
+
+
+class DesiredMachinePool(BaseModel):
+    cluster_name: str
+    hypershift: bool
+    pools: list[ClusterMachinePoolV1]
+
+
+class DesiredStateList(BaseModel):
+    cluster_pools: list[DesiredMachinePool]
+
+
+def fetch_current_state(
+    ocm_map: OCMMap,
+    clusters: Iterable[ClusterV1],
+) -> Mapping[str, list[AbstractPool]]:
+    current_state: MutableMapping[str, list[AbstractPool]] = {}
     for cluster in clusters:
-        cluster_name = cluster["name"]
-        ocm = ocm_map.get(cluster_name)
-        machine_pools = ocm.get_machine_pools(cluster_name)
-        for machine_pool in machine_pools:
-            machine_pool["cluster"] = cluster_name
-            current_state.append(machine_pool)
+        current_state[cluster.name] = []
+        ocm = ocm_map.get(cluster.name)
+        if cluster.spec and cluster.spec.hypershift:
+            machine_pools = ocm.get_node_pools(cluster.name)
+            for machine_pool in machine_pools:
+                current_state[cluster.name].append(
+                    NodePool(
+                        id=machine_pool["id"],
+                        replicas=machine_pool["replicas"],
+                        aws_node_pool=AWSNodePool(
+                            instance_type=machine_pool["aws_node_pool"]["instance_type"]
+                        ),
+                        taints=machine_pool.get("taints"),
+                        labels=machine_pool.get("labels"),
+                        subnet=machine_pool.get("subnet"),
+                        cluster=cluster.name,
+                    )
+                )
+        else:
+            machine_pools = ocm.get_machine_pools(cluster.name)
+            for machine_pool in machine_pools:
+                current_state[cluster.name].append(
+                    MachinePool(
+                        id=machine_pool["id"],
+                        replicas=machine_pool["replicas"],
+                        instance_type=machine_pool["instance_type"],
+                        taints=machine_pool.get("taints"),
+                        labels=machine_pool.get("labels"),
+                        cluster=cluster.name,
+                    )
+                )
 
-    return ocm_map, current_state
+    return current_state
```

Comment:
This is quite long and nested, can be extracted like this to avoid mutation:

```python
def fetch_current_state(
    ocm_map: OCMMap,
    clusters: Iterable[ClusterV1],
) -> Mapping[str, list[AbstractPool]]:
    return {c.name: fetch_current_state_for_cluster(c, ocm_map.get(c.name)) for c in clusters}


def fetch_current_state_for_cluster(cluster, ocm):
    if cluster.spec and cluster.spec.hypershift:
        return [
            NodePool(
                id=machine_pool["id"],
                replicas=machine_pool["replicas"],
                aws_node_pool=AWSNodePool(
                    instance_type=machine_pool["aws_node_pool"]["instance_type"]
                ),
                taints=machine_pool.get("taints"),
                labels=machine_pool.get("labels"),
                subnet=machine_pool.get("subnet"),
                cluster=cluster.name,
            )
            for machine_pool in ocm.get_node_pools(cluster.name)
        ]
    return [
        MachinePool(
            id=machine_pool["id"],
            replicas=machine_pool["replicas"],
            instance_type=machine_pool["instance_type"],
            taints=machine_pool.get("taints"),
            labels=machine_pool.get("labels"),
            cluster=cluster.name,
        )
        for machine_pool in ocm.get_machine_pools(cluster.name)
    ]
```

Reply:
Great refactoring, added!